File: FullTextSearch/src/Utility/QueryProcessor.java
package Utility;

import Model.Document;
import Model.InvertedIndex;
import SetOperators.AndSetOperator;
import SetOperators.OrSetOperator;
import SetOperators.SubtractSetOperator;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class QueryProcessor {

    private List<Document> documents;
    private InvertedIndex invertedIndex;

    public QueryProcessor(final List<Document> documents) {
        setDocuments(documents);
        final InvertedIndex invertedIndex = new InvertedIndex();
        invertedIndex.addWordsOfMultipleDocuments(getDocuments());
        setInvertedIndex(invertedIndex);
    }

    public List<String> advancedQuery(final String input) {

        final InputProcessor inputProcessor = new InputProcessor(input);
        final List<String> orStrings = inputProcessor.getOrStrings();
        final List<String> andStrings = inputProcessor.getAndStrings();
        final List<String> subtractStrings = inputProcessor.getSubtractString();
        Set<String> result = new HashSet<>();

        if (andStrings != null && !andStrings.isEmpty()) {
            result = getInvertedIndex().getIndices().get(andStrings.get(0));
            // because if we have multiple ands, the first one must not be empty.
        }
        if (result == null) {
            result = new HashSet<>();
        }

        // Operation Precedence: AND (Without Signs) -> Or (+ Signs) ->  Subtract (- Signs)
        result= performAllOperations(result, andStrings, orStrings, subtractStrings);

        return new ArrayList<>(result);
    }

    private Set<String> performAllOperations(Set<String> result, List<String> andStrings, List<String> orStrings, List<String> subtractStrings) {
        result = new AndSetOperator().operate(result, andStrings, getInvertedIndex());
        result = new OrSetOperator().operate(result, orStrings, getInvertedIndex());
        result = new SubtractSetOperator().operate(result, subtractStrings, getInvertedIndex());
        return result;
    }


    private List<Document> getDocuments() {
        return documents;
    }

    private void setDocuments(final List<Document> documents) {
        this.documents = documents;
    }

    private InvertedIndex getInvertedIndex() {
        return invertedIndex;
    }

    private void setInvertedIndex(final InvertedIndex invertedIndex) {
        this.invertedIndex = invertedIndex;
    }

}

File: FullTextSearchCsharp/FullTextSearch/FullTextSearch/QueryProcessors/QueryProcessor.cs
using System.Collections.Generic;
using System.Linq;
using FullTextSearch.Model;
using FullTextSearch.SetOperators;
using FullTextSearch.Utility.StringProcessors;
using FullTextSearch.Utility.StringProcessors.InputProcessors;

namespace FullTextSearch.QueryProcessors
{
    public class QueryProcessor : IQueryProcessor
    {
        public InvertedIndex InvertedIndexProvider { get; set; }
        public IInputProcessor InputProcessorProvider { get; set; }

        /*We have two constructors because we are developing a library
        users can simply pass documents to query processor
        Or if they made the inverted Index and inputProcessor themselves
        And just want to use simple searching functionality
        they can just pass them without bothering with documents. */
        public QueryProcessor(List<Document> documents)
        {
            new Tokenizer().TokenizeAllDocuments(documents);
            InputProcessorProvider = new InputProcessor();
            InvertedIndexProvider = new InvertedIndex();
            InvertedIndexProvider.AddWordsOfMultipleDocuments(documents);
        }

        public QueryProcessor(InvertedIndex invertedIndex,IInputProcessor inputProcessor )
        {

            InputProcessorProvider = inputProcessor;
            InvertedIndexProvider = invertedIndex;
        }

        public List<string> PerformSearch(string input)
        {
            InputProcessorProvider.ProcessInput(input);
            var result = PrepareResultSet();
            return DoSearchStrategies(result);
        }

        private ISet<string> PrepareResultSet()
        {
            /*This method is used so if we have any AND string, it must be assigned to result set.
              Because AND has higher precedence and operating it on empty non initialized don't give us appropriate results.  */
            ISet<string> result = new HashSet<string>(); 
            if (InputProcessorProvider.AndStrings.Any())
            {
                var firstWordInAndStrings = InputProcessorProvider.AndStrings[0];
                result.UnionWith(InvertedIndexProvider.Indexes[firstWordInAndStrings]);
            }
            return result;
        }

        private List<string> DoSearchStrategies(ISet<string> result)
        {
            // Precedence: AND -> OR -> SUBTRACT
            result = new AndSetOperator().Operate(result, InputProcessorProvider.AndStrings, InvertedIndexProvider);
            result = new OrSetOperator().Operate(result, InputProcessorProvider.OrStrings, InvertedIndexProvider);
            result = new SubtractSetOperator().Operate(result, InputProcessorProvider.SubtractStrings, InvertedIndexProvider);
            return result.ToList().OrderBy(i => i).ToList();
        }
    }
}